src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,141 @@
+        return snsMessage != null ?
+            ExtractParentContext(snsMessage) :
+            Propagators.DefaultTextMapPropagator.Extract(default, sqsMessage.MessageAttributes, SqsMessageAttributeGetter);
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent snsEvent)
+    {
+        if (snsEvent == null)
+        {
+            return default;
+        }
+
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var record = snsEvent.Records.LastOrDefault();
+        return ExtractParentContext(record);
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSRecord record)
+    {
+        return (record?.Sns != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, record.Sns.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    internal static PropagationContext ExtractParentContext(SNSEvent.SNSMessage message)
+    {
+        return (message != null) ?
+            Propagators.DefaultTextMapPropagator.Extract(default, message.MessageAttributes, SnsMessageAttributeGetter) :
+            default;
+    }
+
+    private static IEnumerable<string> SqsMessageAttributeGetter(IDictionary<string, SQSEvent.MessageAttribute> attributes, string attributeName)
+    {
+        SQSEvent.MessageAttribute attribute = attributes.GetValueByKeyIgnoringCase(attributeName);
+        if (attribute == null)
+        {
+            return null;
+        }
+
+        return attribute.StringValue != null ?
+            new[] { attribute.StringValue } :
+            attribute.StringListValues;
+    }
+
+    private static IEnumerable<string> SnsMessageAttributeGetter(IDictionary<string, SNSEvent.MessageAttribute> attributes, string attributeName)
+    {
+        SNSEvent.MessageAttribute attribute = attributes.GetValueByKeyIgnoringCase(attributeName);
+        if (attribute == null)
+        {
+            return null;
+        }
+
+        switch (attribute.Type)
+        {
+            case SnsAttributeTypeString when attribute.Value != null:
+                return new[] { attribute.Value };
+            case SnsAttributeTypeStringArray when attribute.Value != null:
+                // Multiple values are stored as CSV (https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html).
+                return attribute.Value.Split(',');
+            default:
+                return null;
+        }
+    }
+
+    private static SNSEvent.SNSMessage GetSnsMessage(SQSEvent.SQSMessage sqsMessage)
+    {
+        SNSEvent.SNSMessage snsMessage = null;
+
+        var body = sqsMessage.Body;
+        if (body != null &&
+            body.TrimStart().StartsWith("{") &&
+            body.Contains(SnsMessageAttributes))
+        {
+            try
+            {
+                snsMessage = JsonConvert.DeserializeObject<SNSEvent.SNSMessage>(body);
```

Comment:
Instead of deserializing the full to a .NET type, have you considered using lower-level JSON APIs to just deserialize the attributes? I don't know if this would be more efficient though, so in doubt we can leave it as-is.

Reply:
I don't see benefits of just deserializing the attributes. I'm not sure about efficiency 'the whole class'  vs 'only attributes'. Let's leave it as-is.